spec/02-integration/06-invalidations/04-balancer_cache_correctness_spec.lua

```diff
@@ -0,0 +1,39 @@
+local helpers = require "spec.helpers"
+
+local POLL_INTERVAL = 0.3
+
+for _, strategy in helpers.each_strategy() do
+  describe("balancer cache [#" .. strategy .. "]", function()
+
+    lazy_setup(function()
+      -- truncate the database so we have zero upstreams
+      helpers.get_db_utils(strategy, { "upstreams", })
+
+      local db_update_propagation = strategy == "cassandra" and 0.1 or 0
+
+      assert(helpers.start_kong {
+        log_level             = "debug",
+        database              = strategy,
+        proxy_listen          = "0.0.0.0:8000, 0.0.0.0:8443 ssl",
+        admin_listen          = "0.0.0.0:8001",
+        db_update_frequency   = POLL_INTERVAL,
+        db_update_propagation = db_update_propagation,
+        nginx_conf            = "spec/fixtures/custom_nginx.template",
+      })
+    end)
+
+    lazy_teardown(function()
+      helpers.stop_kong()
+    end)
+
+
+    -- https://github.com/Kong/kong/issues/8970
+    it("upstreams won't reload at unusual rate", function()
+      assert.logfile().has.line("loading upstreams dict into memory", true, 5)
+      -- turncate log
+      io.open("./servroot/logs/error.log", "w"):close()
+      ngx.sleep(20)
+      assert.logfile().has.no.line("loading upstreams dict into memory", true)
```

Comment:
```suggestion
      assert.logfile().has.no.line("loading upstreams dict into memory", true, 20)
```

I know it's not very obvious/intuitive, but the timeout is still useful in the negative case. It will continue checking the log file for the full 20s.

Reply:
Alternatively, I wonder if it's possible to cover this with a unit test instead?